beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -0,0 +1,337 @@
+type blocksFetcher struct {
+	ctx                    context.Context
+	headFetcher            blockchain.HeadFetcher
+	p2p                    p2p.P2P
+	rateLimiter            *leakybucket.Collector
+	requests               chan *fetchRequestParams   // incoming fetch requests from downstream clients
+	receivedFetchResponses chan *fetchRequestResponse // responses from peers are forwarded to downstream clients
+	quit                   chan struct{}              // termination notifier
+}
+
+// fetchRequestParams holds parameters necessary to schedule a fetch request.
+type fetchRequestParams struct {
+	start uint64 // starting slot
+	count uint64 // how many slots to receive (fetcher may return fewer slots)
+}
+
+// fetchRequestResponse is a combined type to hold results of both successful executions and errors.
+// Valid usage pattern will be to check whether result's `err` is nil, before using `blocks`.
+type fetchRequestResponse struct {
+	params *fetchRequestParams
+	blocks []*eth.SignedBeaconBlock
+	err    error
+}
+
+// newBlocksFetcher creates ready to use fetcher
+func newBlocksFetcher(cfg *blocksFetcherConfig) *blocksFetcher {
+	return &blocksFetcher{
+		ctx:                    cfg.ctx,
+		headFetcher:            cfg.headFetcher,
+		p2p:                    cfg.p2p,
+		rateLimiter:            cfg.rateLimiter,
+		requests:               make(chan *fetchRequestParams),
+		receivedFetchResponses: make(chan *fetchRequestResponse),
+		quit:                   make(chan struct{}),
+	}
+}
+
+// start boots up the fetcher, which starts listening for incoming fetch requests.
+func (f *blocksFetcher) start() {
+	go f.loop()
+}
+
+// stop terminates all fetcher operations.
+func (f *blocksFetcher) stop() {
+	close(f.quit)
+}
+
+// iter returns an outgoing channel, on which consumers are expected to constantly iterate for results/errors.
+func (f *blocksFetcher) iter() <-chan *fetchRequestResponse {
+	return f.receivedFetchResponses
+}
+
+// loop is a main fetcher loop, listens for incoming requests/cancellations, forwards outgoing responses
+func (f *blocksFetcher) loop() {
+	defer close(f.receivedFetchResponses)
+
+	randGenerator := rand.New(rand.NewSource(time.Now().Unix()))
+	highestFinalizedSlot := helpers.StartSlot(f.highestFinalizedEpoch() + 1)
+	maxPeersToSync := params.BeaconConfig().MaxPeersToSync
+
+	for {
+		select {
+		case <-f.ctx.Done():
+			// Upstream context is done.
+			log.Warn("Upstream context canceled. Blocks fetcher is stopped.")
+			return
+		case <-f.quit:
+			// Terminating abort all operations.
+			return
+		case req := <-f.requests:
+			root, finalizedEpoch, peers := f.p2p.Peers().BestFinalized(maxPeersToSync, helpers.SlotToEpoch(f.headFetcher.HeadSlot()))
+			log.WithFields(logrus.Fields{
+				"request":        req,
+				"finalizedEpoch": finalizedEpoch,
+				"numPeers":       len(peers),
+			}).Debug("Block fetcher receives request")
+
+			if len(peers) == 0 {
```

Comment:
I think this piece of code still exists in the `round_robin` main function. Either have it here or there.